providers/implementations/signature/rsa.c

```diff
@@ -0,0 +1,962 @@
+    int pad_mode;
+    /* message digest for MGF1 */
+    char mgf1_mdname[80];        /* Should be big enough */
+    char mgf1_mdprops[80];       /* Should be big enough */
+    EVP_MD *mgf1_md;
+    /* PSS salt length */
+    int saltlen;
+    /* Minimum salt length or -1 if no PSS parameter restriction */
+    int min_saltlen;
+
+    /* Temp buffer */
+    unsigned char *tbuf;
+
+} PROV_RSA_CTX;
+
+static void *rsa_newctx(void *provctx)
+{
+    PROV_RSA_CTX *prsactx = OPENSSL_zalloc(sizeof(PROV_RSA_CTX));
+
+    if (prsactx == NULL)
+        return NULL;
+
+    prsactx->libctx = PROV_LIBRARY_CONTEXT_OF(provctx);
+    return prsactx;
+}
+
+/* True if PSS parameters are restricted */
+#define rsa_pss_restricted(prsactx) (prsactx->min_saltlen != -1)
+
+static int rsa_signature_init(void *vprsactx, void *vrsa)
+{
+    PROV_RSA_CTX *prsactx = (PROV_RSA_CTX *)vprsactx;
+
+    if (prsactx == NULL || vrsa == NULL || !RSA_up_ref(vrsa))
+        return 0;
+
+    RSA_free(prsactx->rsa);
+    prsactx->rsa = vrsa;
+    if (RSA_get0_pss_params(prsactx->rsa) != NULL)
+        prsactx->pad_mode = RSA_PKCS1_PSS_PADDING;
+    else
+        prsactx->pad_mode = RSA_PKCS1_PADDING;
+    /* Maximum for sign, auto for verify */
+    prsactx->saltlen = RSA_PSS_SALTLEN_AUTO;
+    prsactx->min_saltlen = -1;
+
+    return 1;
+}
+
+static int setup_md(PROV_RSA_CTX *ctx)
+{
+    if (ctx->mdname[0] != '\0') {
+        EVP_MD *md = EVP_MD_fetch(ctx->libctx, ctx->mdname, ctx->mdprops);
+
+        if (md == NULL)
+            return 0;
+        EVP_MD_CTX_free(ctx->mdctx);
+        EVP_MD_free(ctx->md);
+        ctx->md = md;
+        ctx->mdsize = EVP_MD_size(md);
+        ctx->mdctx = EVP_MD_CTX_new();
+        if (ctx->mdctx == NULL)
+            return 0;
+
+        if (!EVP_DigestInit_ex(ctx->mdctx, md, NULL))
+            return 0;
+    }
+
+    if (ctx->mgf1_mdname[0] != '\0') {
+        EVP_MD *md = EVP_MD_fetch(ctx->libctx, ctx->mgf1_mdname,
+                                  ctx->mgf1_mdprops);
+
+        if (md == NULL)
+            return 0;
+        EVP_MD_free(ctx->mgf1_md);
+        ctx->mgf1_md = md;
+    }
+
+    return 1;
+}
+
+static int setup_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        return 1;
+    if ((ctx->tbuf = OPENSSL_malloc(RSA_size(ctx->rsa))) == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    return 1;
+}
+
+static void clean_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        OPENSSL_cleanse(ctx->tbuf, RSA_size(ctx->rsa));
+}
+
+static void free_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        OPENSSL_clear_free(ctx->tbuf, RSA_size(ctx->rsa));
+}
+
+static int rsa_sign(void *vprsactx, unsigned char *sig, size_t *siglen,
+                    size_t sigsize, const unsigned char *tbs, size_t tbslen)
+{
+    PROV_RSA_CTX *prsactx = (PROV_RSA_CTX *)vprsactx;
+    int ret;
+
+    if (sig == NULL) {
+        *siglen = RSA_size(prsactx->rsa);
```

Comment:
This might not be quite right...

Reply:
rsa_sign.pod thinks it is correct??

Reply:
Yeah, sorry, got mixed up. RSA encodes a X509_SIG, *then* public-encrypts the result, while DSA does it the other way around...